proto/proto/eraftpb.proto

```diff
@@ -94,31 +87,94 @@ message HardState {
     uint64 commit = 3;
 }
 
+enum ConfChangeTransition {
+    // Automatically use the simple protocol if possible, otherwise fall back
+    // to ConfChangeJointImplicit. Most applications will want to use this.
+    Auto = 0;
+    // Use joint consensus unconditionally, and transition out of them
+    // automatically (by proposing a zero configuration change).
+    //
+    // This option is suitable for applications that want to minimize the time
+    // spent in the joint configuration and do not store the joint configuration
+    // in the state machine (outside of InitialState).
+    Implicit = 1;
+    // Use joint consensus and remain in the joint configuration until the
+    // application proposes a no-op configuration change. This is suitable for
+    // applications that want to explicitly control the transitions, for example
+    // to use a custom payload (via the Context field).
+    Explicit = 2;
+}
+
 message ConfState {
-    repeated uint64 nodes = 1;
+    repeated uint64 voters = 1;
     repeated uint64 learners = 2;
+
+    // The voters in the outgoing config. It's not empty means it's in joint consensus.
+    repeated uint64 voters_outgoing = 3;
+    // The nodes that will become learners when the outgoing config is removed.
+    // These nodes are necessarily currently in nodes_joint (or they would have
+    // been added to the incoming config right away).
+    repeated uint64 learners_next = 4;
+    // If set, the config is joint and Raft will automatically transition into
+    // the final config (i.e. remove the outgoing config) when this is safe.
+    bool auto_leave = 5;
 }
 
 enum ConfChangeType {
     AddNode    = 0;
     RemoveNode = 1;
     AddLearnerNode = 2;
-    BeginMembershipChange = 3;
-    FinalizeMembershipChange = 4;
 }
 
 message ConfChange {
-    uint64 id = 1;
     ConfChangeType change_type = 2;
-    // Used in `AddNode`, `RemoveNode`, and `AddLearnerNode`.
     uint64 node_id = 3;
-
-    // Some extra information used in configuration change.
     bytes context = 4;
 
-    // For `BeginMembershipChange`, the target configuration is stored in the field.
-    ConfState configuration = 5;
+    uint64 id = 1;
```

Comment:
Why this move?

Reply:
Just to keep same order with etcd/raft. In fact the field `id` is only used by etcd, maybe we can remove it later.